doc/sphinx/source/styleguide.rst

```diff
@@ -1,4 +1,455 @@
+so that their data members are visible and accessible. Examples of structs in
+the TDycore library are the ``MaterialProp`` and ``CharacteristicCurve`` types,
+which represent sets of material properties and parameterized curves (a
+saturation curve, for example), respectively.
+
+Since a struct is a simple container without behavior (a "Plain Old Datatype",
+or "POD"), no associated functions are needed. However, such functions can be
+provided if they make the struct/POD more convenient to use.
+
+Classes
+-------
+
+A "class" in TDycore is a struct whose definition is **private**--not exposed
+within the public interface defined by the set of header files intended for use
+by external developers. This means that a developer cannot directly manipulate
+the fields within a "class". Instead, one manipulates a class using its
+**interface**--a set of functions associated with that class.
+
+Define class bodies in source files only, unless their internal structure is
+intended to be explicitly exposed to developers. "Typedef" your class type so
+the ``struct`` keyword can be omitted from its type.
+
+The struct and functions defining a class are governed by a few simple
+conventions described below. PETSc uses similar conventions.
+
+Class Type (Struct)
+^^^^^^^^^^^^^^^^^^^
+
+The struct representing the class type has the same name as the class itself.
+The type of the class itself is a pointer to its underlying struct. The
+underlying struct is declared with a ``_p`` suffix. Then the class is declared
+as a ``typedef`` to a pointer to the struct type. For example, if you want to
+declare a "washing machine" class, first declare its underlying struct (without
+defining its body):::
+
+    typedef struct WashingMachine_p;
+
+in an appropriate header file. Then define the class itself:::
+
+    typedef WashingMachine_p* WashingMachine;
+
+Class Constructor(s)
+^^^^^^^^^^^^^^^^^^^^
+
+Typically, a class has a single constructor function named after the class, with
+``New`` at the end. A constructor takes a number of arguments for initializing
+the class, plus a final argument that stores a pointer to a newly-allocated
+instance of the class. For example, consider the following constructor for our
+``WashingMachine`` class:::
+
+    PetscErrorCode WashingMachineNew(PetscInt numCents, WashingMachine* wm);
+
+This constructor creates a ``WashingMachine`` instance that costs the given
+number of cents to wash a load of laundry. The ``wm`` argument stores the
+new instance. The constructor returns an integer-valued error code described
+in the section on functions.
+
+Sometimes it's convenient to provide more than one constructor, or a
+constructor that converts another datatype to a given instance of a class.
+In these cases, name each constructor so that it briefly conveys its purpose.
+For example, a constructor that converts an array of ``real_t`` to a point
+might be declared
+
+``point_t* point_from_array(real_t* array);``
```

Comment:
I don't know others' opinions, but PETSc intentionally doesn't name parameters in declarations because the compiler doesn't check them and they can become stale/misleading.

I don't care in function declarations, but inside functions, I prefer
```c
real_t *a, b;
```
over
```c
real_t* a, b;
```
since the latter is often misread is suggesting that `b` would have type `real_t*`.
Style guides that recommend the `real_t* a` spacing often say only one variable can be declared per line, though that eats up precious vertical space on our luxuriously wide displays.

Reply:
Right on. I'll update the guide to reflect this.